---
title: "Assignment9_jek4zy"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Biobase)
library(GSE5859Subset)
data(GSE5859Subset)
library(qvalue)
library(genefilter)
#source("https://bioconductor.org/biocLite.R")
#biocLite("sva")
library(sva)
```

###Exercises on page 419

```{r}
admissions <- read.csv("admissions.csv")
admissions
```

1. What is the proportion of women that were accepted?
```{r}
index = which(admissions$Gender==1)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```
Answer: Men: 0.4451951, Women: 0.3033351

2. Now that we have observed different acceptance rates between genders, test for the
significance of this result. If you perform an independence test, what is the p-value?
```{r}
a<-as.data.frame(matrix(0,2,2))
rownames(a)<-c("Men","Women")
colnames(a)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
a[1,1]= sum(men$Number * men$Percent/100)
a[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
a[2,1]= sum(women$Number * women$Percent/100)
a[2,2]= sum(women$Number * (1-women$Percent/100))
chisq.test(a)$p.value
```
Answer: p-value = 9.14*10^-22

3. We can quantify how “hard” a major is by using the percent of students that were accepted.
Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?
```{r}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
major[which.min(H)]
min(H)
```
Answer: Major F

4. What proportion is accepted for this major?
Answer: 0.06477591 (see code above)

5. For men, what is the correlation between the number of applications across majors and H?
```{r}
cor(men$Number,H)
```
Answer: 0.7647567

6. For women, what is the correlation between the number of applications across majors and
H?
```{r}
cor(women$Number,H)
```
Answer: -0.6743393

7. Given the answers to the above, which best explains the differences in admission percentages
when we combine majors?
Answer: C. The difference in admission percentages are best explained by the correlation between gender and the "hardness" level of intended major. Since more women applied to harder majors and more of the men that applied choose "easier" majors, when we combine majors we see an overall discrepency in admission percentages. 

###Exercises on page 431
```{r}
library(Biobase)
library(GSE5859)
data(GSE5859)
geneExpression = exprs(e)
sampleInfo = pData(e)
```

1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here.
Note that ethnic group and year is almost perfectly confounded:
```{r}
sampleInfo
year = format(sampleInfo$date,"%y")
table(year,sampleInfo$ethnicity)
```

2. For how many of these years do we have more than one ethnicity represented?
```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```
Answer: 2 (year 05 and 06)

3. Repeat the above exercise, but now, instead of year, consider the month as well. Specifically, instead of the year variable defined above use: 
month.year = format(sampleInfo$date,"%m%y")
For what proportion of these month.year values do we have more than one ethnicity represented?
```{r}
month.year = format(sampleInfo$date,"%m%y")
tab = table(month.year,sampleInfo$ethnicity)
print(tab)
x = rowSums( tab != 0)
mean( x >= 2)
```
Answer: 0.04761905

4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ?
```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Answer: 4308

5. What is the estimate of pi0 provided by qvalue:
```{r}
qval$pi0
```
Answer:0.3628642

6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those
processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many
genes have q-values less than 0.05?
```{r}
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Answer: 2463

7. Now we are going to compare ethnicities as was done in the original publication in which
these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values.
How many genes have q-values < 0.05 ?
```{r}
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Answer: 7217

8. Over 80% of genes are called differentially expressed between ethnic groups. However, due
to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the “major effect” in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005.
How many genes have q-values < 0.05 ?
Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Answer: 560

9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3)
How many genes have q-values < 0.05 ?
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
Answer: 3695

###Exercises on page 442

```{r}
library(GSE5859Subset)
data(GSE5859Subset)
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
table( sampleInfo$group, month)
```

1. Using the functions rowt tests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%.
```{r}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.1)
```
Answer: 59

2. For the list calculated above, what proportion of this list is on chrX or chrY?
```{r}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
Answer: 0.3389831

3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?
```{r}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = which(qvals<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
month = factor( format(sampleInfo$date,"%m"))
pval = rowttests(geneExpression[index,], month)$p.value
mean(pval<0.05)
```
Answer: 0.8717849

4. What proportion of these have p-values <0.05 ?
```{r}
filter <- geneAnnotation[geneAnnotation$CHR != "chrX" & geneAnnotation$CHR != "chrY" & qval$qvalues < 0.1, ]
autosomal <- geneExpression[filter$PROBEID,]
month = factor( format(sampleInfo$date,"%m"))
ttxt2 <- rowttests(autosomal, month)
mean(ttxt2$p.val < 0.05)
```
Answer: 0.8717949

5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model:
Which of the following creates the appropriate design matrix? 
• A)X=model.matrix(∼sex+ethnicity)
• B)X=cbind(sex,as.numeric(month))
• C) It can’t be done with one line.
• D)X=model.matrix(∼sex+month)

Answer: D

6. How many of the q-values for the group comparison are now <0.1?
```{r}
library(limma)
X <- model.matrix(~sex+month)
fit <- lmFit(geneExpression,X)
k <- 2
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
Answer: 17

7. With this new list, what proportion of these are chrX and chrY?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
Answer: 0.8823529

8. How many on Y or X?
```{r}
mean(index) * 17
```
Answer: 15

9. Now from the linear model above, extract the p-values related to the coefficient representing
the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1?
```{r}
library(limma)
X <- model.matrix(~sex+month)
fit <- lmFit(geneExpression,X)
k <- 3 ##third coef, for month
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
Answer: 3170

###Exercises on page 448

```{r}
library(Biobase)
library(GSE5859Subset)
data(GSE5859Subset)
```

1. Suppose you want to make an MA plot of the first two samples y= geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?
Answer: B

2. Say Y is M×N,in theS VDY = UDV⊤ which of the following is not correct? 
• A) DV ⊤ are the new coordinates for the projection U⊤Y
• B) UD are the new coordinates for the projection YV
• C) D are the coordinates of the projection U⊤Y
• D) U⊤Y is a projection from an N-dimensional to M-dimensional subspace.
Answer: C

3. Define: y = geneExpression - rowMeans(geneExpression)
Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted.
```{r}
y = geneExpression - rowMeans(geneExpression)
library(rafalib)
sex = sampleInfo$group
mypar(1,2)
cors = cor(y)
image(cors)
o = order(sampleInfo$date)
image(cors[o,o])
```

Based on these plots, which of the following you would say is true?
• A) The samples appear to be completely independent of each other.
• B) Sex seems to be creating structures as evidenced by the two cluster of highly correlated samples.
• C) There appear to be only two factors completely driven by month.
• D) The fact that in the plot ordered by month we see two groups mainly driven by month, and within these we see subgroups driven by date, seems to suggest date more than month per se are the hidden factors.

Answer: Definitely not A, the samples definitely have some confounding factors. C is not wrong but I am more convinced by D, because the plots show that main groups are mainly driven by date order, of which month may play a larger role.

4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates. 
```{r}
s <- svd(y)
What <- s$v[,1:2] 
rownames(What)<-colnames(y)
round(What,2)
fit = s$u[,1:2]%*% (s$d[1:2]*t(What))
var(as.vector(fit))/var(as.vector(y))
```
Which command gives us these estimates?
• A)pcs=svd(y)$v[1:2,]
• B)pcs=svd(y)$v[,1:2]
• C)pcs=svd(y)$u[,1:2]
• D)pcs=svd(y)$d[1:2]

Answer: B) pcs=svd(y)$v[,1:2]

5. Plot each of the estimated factors ordered by date. Use color to denote month. The first factor is clearly related to date.
```{r}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)
}
```
Which of the following appear to be most different according to this factor?
• A)June23andJune27 • B)Oct07andOct28
• C)June10andJune23 • D)June15andJune24

Answer: A) June 23rd and June 27th

6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?
```{r}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```
Answer: 2

7. Which PC most correlates (negative or positive correlation) with month?
```{r}
s = svd(y)
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
plot(t(cors))
which.max(abs(cors))
```
Answer: January (month 1)

8. What is this correlation (in absolute value)?
```{r}
max(abs(cors))
```
Answer: 0.8297915

9. Which PC most correlates (negative or positive correlation) with sex?
```{r}
s = svd(y)
sex = sampleInfo$group
cors = cor( as.numeric(sex),s$v)
plot(t(cors))
which.max(abs(cors))
```
Answer: January 

10. What is this correlation (in absolute value)?
```{r}
max(abs(cors))
```
Answer: 0.6236858

11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?
```{r}
X <- model.matrix(~sex+s$v[,1:2])
fit <- lmFit(y,X)
k <- 2 
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
Answer: 14

12. What proportion of the genes are on chromosomes X and Y?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
Answer: 1 (all of them)

###Exercises on page 459

1. In a previous section we estimated factors using PCA, but we noted that the first factor was correlated with our outcome of interest:
```{r}
    s <- svd(geneExpression-rowMeans(geneExpression))
    cor(sampleInfo$group,s$v[,1])
```
The svafit function estimates factors, but downweighs the genes that appear to correlate with the outcome of interest. It also tries to estimate the number of factors and returns the estimated factors like this:
```{r}
    sex = sampleInfo$group
    mod = model.matrix(~sex)
    svafit = sva(geneExpression,mod)
    head(svafit$sv)
```
The resulting estimated factors are not that different from the PCs.
```{r}
for(i in 1:ncol(svafit$sv)){
print( cor(s$v[,i],svafit$sv[,i]) ) }
```
Now fit a linear model to each gene that instead of month includes these factors in the model. Use the qvalue function. How many genes have q-value < 0.1?
```{r}
svaX<-model.matrix(~sex+svafit$sv)
fit <- lmFit(geneExpression,svaX)
k <- 2
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
Answer: 13

2. How many of these genes are from chrY or chrX?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
mean(index) * 13
```
Answer: 12 (or .9230769 of the genes are in chrY or chrX)